# 4.4 Data Wrangling & Subsetting

In [88]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [93]:
# Reletive path
path = '../'

# Loading dataframe orders.csv
df_ords = pd.read_csv(
    os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), 
    index_col = False
)

In [90]:
# Import the “products.csv
df_prods = pd.read_csv(
    os.path.join(path, '02 Data', 'Original Data', 'products.csv'), 
    index_col = False
)

In [95]:
# df_ords = df_ords.drop(columns = ['eval_set'])
# df_ords['variable'].value_counts(dropna = False)
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

df_ords['order_id'] = df_ords['order_id'].astype('str')
# df_ords['order_id'].dtype()

In [96]:
# Loading dataframe
df_dep = pd.read_csv(
    os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), 
    index_col = False
)

df_dep.head(5)

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [97]:
# Transposing
df_dep_t = df_dep.T

df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [98]:
# Take the first row
new_header = df_dep_t.iloc[0]

In [99]:
# Deleting first row
df_dep_t_new = df_dep_t[1:]

In [100]:
# Set the new header row as the df header
df_dep_t_new.columns = new_header

In [101]:
# Set the new header row as the df header
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [102]:
# Converting to dict
data_dict = df_dep_t_new.to_dict('index')

In [103]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [104]:
print(data_dict.get('19'))

{'department': 'snacks'}


In [105]:
# Sorting examples
df_snacks =  df_prods[df_prods['department_id']==19]
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([17,18,19])]

In [106]:
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

# Task 4.4 exercises

In [107]:
# 2. Find another identifier variable in the df_ords dataframe that doesn’t need to be included 
#    in your analysis as a numeric variable and change it to a suitable format.

df_ords[['order_id', 'user_id']] = df_ords[['order_id', 'user_id']].astype('str')

In [108]:
df_ords

order_id user_id eval_set  order_number  orders_day_of_week  \
0        2539329       1    prior             1                   2   
1        2398795       1    prior             2                   3   
2         473747       1    prior             3                   3   
3        2254736       1    prior             4                   4   
4         431534       1    prior             5                   4   
...          ...     ...      ...           ...                 ...   
3421078  2266710  206209    prior            10                   5   
3421079  1854736  206209    prior            11                   4   
3421080   626363  206209    prior            12                   1   
3421081  2977660  206209    prior            13                   1   
3421082   272231  206209    train            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [109]:
# 3. Look for a variable in your df_ords dataframe with an unintuitive name and change its name 
#    without overwriting the data frame.

df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [110]:
# 4. Your client wants to know what the busiest hour is for placing orders. 
#    Find the frequency of the corresponding variable and share your findings.

df_ords['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

The busiest hour for palcing orders is 10 am (288418)

In [111]:
# 5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods 
#    dataframe using a data dictionary.

print("The meaning behind a value of 4 is:", data_dict.get('4')['department'])

The meaning behind a value of 4 is: produce


In [112]:
# 6. The sales team in your client’s organization wants to know more about breakfast item sales. 
#    Create a subset containing only the required information.

df_breakfast =  df_prods[df_prods['department_id']==14]
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [113]:
# 7. They’d also like to see details about products that customers might use to throw dinner parties. 
# Your task is to find all observations from the entire dataframe that include items 
# from the following departments: alcohol, deli, beverages, and meat/seafood. 
# You’ll need to present this subset to your client.

df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5, 7, 12, 20])]
df_dinner_parties

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

In [114]:
# 8. It’s important that you keep track of total counts in your dataframes. 
#    How many rows does the last dataframe you created have?

df_dinner_parties.shape[0]

7650

In [115]:
# 9. Someone from the data engineers team in Instacart thinks they’ve spotted 
#    something strange about the customer with a "user_id" of “1.” 
#    Extract all the information you can about this user.

user_1_orders = df_ords[df_ords['user_id']=='1']
user_1_orders

order_id user_id eval_set  order_number  orders_day_of_week  \
0   2539329       1    prior             1                   2   
1   2398795       1    prior             2                   3   
2    473747       1    prior             3                   3   
3   2254736       1    prior             4                   4   
4    431534       1    prior             5                   4   
5   3367565       1    prior             6                   2   
6    550135       1    prior             7                   1   
7   3108588       1    prior             8                   1   
8   2295261       1    prior             9                   1   
9   2550362       1    prior            10                   4   
10  1187899       1    train            11                   4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

In [116]:
# 10. You also need to provide some details about this user’s behavior. 
#     What basic stats can you provide based on the information you have?

user_1_orders.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

Main observations:
1) Usually user make orders not less then in 2 weeks, but the 9-th order were made the same day as 8-th.

2) Typically, user place orders during working hours on working days.

In [117]:
# Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

# Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder 
# so that you have a “.csv” file of your departments data in the correct format.
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))